drivers/i2c/i2c_sam4l_twim.c

```diff
@@ -0,0 +1,641 @@
+	} else {
+		twim->CWGR = cwgr_reg_val;
+	}
+
+	LOG_DBG("per_clk: %d, f_prescaled: %d, cwgr_exp: 0x%02x,"
+		"cwgr_reg_val: 0x%08x", per_clk, f_prescaled,
+		cwgr_exp, cwgr_reg_val);
+
+	/* Set clock and data slew rate */
+	twim->SRR = ((speed == I2C_SPEED_FAST_PLUS) ? TWIM_SRR_FILTER(2) :
+						      TWIM_SRR_FILTER(3)) |
+		     TWIM_SRR_CLSLEW(cfg->std_clk_slew_lim) |
+		     TWIM_SRR_CLDRIVEL(cfg->std_clk_strength_low) |
+		     TWIM_SRR_DASLEW(cfg->std_data_slew_lim) |
+		     TWIM_SRR_DADRIVEL(cfg->std_data_strength_low);
+
+	twim->HSSRR = TWIM_HSSRR_FILTER(1) |
+		      TWIM_HSSRR_CLSLEW(cfg->hs_clk_slew_lim) |
+		      TWIM_HSSRR_CLDRIVEH(cfg->hs_clk_strength_high) |
+		      TWIM_HSSRR_CLDRIVEL(cfg->hs_clk_strength_low) |
+		      TWIM_HSSRR_DASLEW(cfg->hs_data_slew_lim) |
+		      TWIM_HSSRR_DADRIVEL(cfg->hs_data_strength_low);
+
+	return 0;
+}
+
+static int i2c_sam_twim_configure(const struct device *dev, uint32_t config)
+{
+	uint32_t bitrate;
+	int ret;
+
+	if (!(config & I2C_MODE_MASTER)) {
+		LOG_ERR("Master Mode is not enabled");
+		return -EIO;
+	}
+
+	if (config & I2C_ADDR_10_BITS) {
+		LOG_ERR("I2C 10-bit addressing is currently not supported");
+		LOG_ERR("Please submit a patch");
+		return -EIO;
+	}
+
+	/* Configure clock */
+	switch (I2C_SPEED_GET(config)) {
+	case I2C_SPEED_STANDARD:
+		bitrate = BUS_SPEED_STANDARD_HZ;
+		break;
+	case I2C_SPEED_FAST:
+		bitrate = BUS_SPEED_FAST_HZ;
+		break;
+	case I2C_SPEED_FAST_PLUS:
+		bitrate = BUS_SPEED_PLUS_HZ;
+		break;
+	case I2C_SPEED_HIGH:
+		LOG_ERR("API not support High Speed value");
+		return -EIO;
+	default:
+		LOG_ERR("Unsupported I2C speed value");
+		return -EIO;
+	}
+
+	/* Setup clock waveform */
+	ret = i2c_clk_set(dev, bitrate);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static void i2c_prepare_xfer_data(struct i2c_sam_twim_dev_data *data)
+{
+	struct i2c_msg *next_msg = NULL;
+
+	if (data->next_nb_bytes > TWIM_MAX_NBYTES_PER_XFER) {
+		data->cur_remaining = TWIM_MAX_NBYTES_PER_XFER;
+
+		data->next_nb_bytes -= TWIM_MAX_NBYTES_PER_XFER;
+		data->next_is_valid = true;
+		data->next_need_rs = false;
+	} else {
+		data->cur_remaining = data->next_nb_bytes;
+
+		if ((data->msg_next_idx + 1) < data->msg_max_idx) {
+			next_msg = &data->msgs[++data->msg_next_idx];
+
+			data->next_nb_bytes = next_msg->len;
+			data->next_is_valid = true;
+			data->next_need_rs = true;
+		} else {
+			data->next_nb_bytes = 0;
+			data->next_is_valid = false;
+			data->next_need_rs = false;
+		}
+	}
+}
+
+static uint32_t i2c_prepare_xfer_cmd(struct i2c_sam_twim_dev_data *data,
+				     uint32_t *cmdr_reg,
+				     uint32_t next_msg_idx)
+{
+	struct i2c_msg *next_msg = &data->msgs[next_msg_idx];
+	bool next_msg_is_read;
+	uint32_t next_nb_remaining;
+
+	*cmdr_reg &= ~(TWIM_CMDR_NBYTES_Msk |
+		       TWIM_CMDR_ACKLAST |
+		       TWIM_CMDR_START |
+		       TWIM_CMDR_READ);
+
+	next_msg_is_read = (next_msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ;
```

Comment:
OK, seeing it on one line I agree parentheses are justifiable and I might well have done that too, because of the `x = y == z;` structure.  I wouldn't have done it for `x = y op z;` for any other operator.  (I think it's that I'd have broken the line before the `==`; with the operator at the end of the line it was harder to immediately perceive what was going on.)

Reply:
I understood your point. But I kept it as original version to be consistent with all others line breaks.